src/wallet/wallet_tx_builder.cpp

```diff
@@ -80,45 +85,48 @@ WalletTxBuilder::GetChangeAddress(
 
     auto changeAddressForZUFVK = [&](
             const ZcashdUnifiedFullViewingKey& zufvk,
-            const std::set<ReceiverType>& receiverTypes) {
+            const std::set<ReceiverType>& receiverTypes)
+        -> tl::expected<ChangeAddress, AddressResolutionError> {
         auto addr = zufvk.GetChangeAddress(getAllowedChangePools(receiverTypes));
         assert(addr.has_value());
         return addr.value();
     };
 
     return examine(selector.GetPattern(), match {
-        [&](const CKeyID&) -> ChangeAddress {
+        [&](const CKeyID&) {
             return changeAddressForTransparentSelector({ReceiverType::P2PKH});
         },
-        [&](const CScriptID&) -> ChangeAddress {
+        [&](const CScriptID&) {
             return changeAddressForTransparentSelector({ReceiverType::P2SH});
         },
-        [](const libzcash::SproutPaymentAddress& addr) -> ChangeAddress {
+        [](const libzcash::SproutPaymentAddress& addr)
+            -> tl::expected<ChangeAddress, AddressResolutionError> {
             // for Sprout, we return change to the originating address using the tx builder.
             return addr;
         },
-        [](const libzcash::SproutViewingKey& vk) -> ChangeAddress {
+        [](const libzcash::SproutViewingKey& vk)
+            -> tl::expected<ChangeAddress, AddressResolutionError> {
             // for Sprout, we return change to the originating address using the tx builder.
             return vk.address();
         },
-        [&](const libzcash::SaplingPaymentAddress& addr) -> ChangeAddress {
+        [&](const libzcash::SaplingPaymentAddress& addr) {
```

Comment:
It's better to have to annotate the return type as `tl::expected<ChangeAddress, AddressResolutionError>` here and below, than to do so in `changeAddressForSaplingAddress` and `changeAddressForZUFVK`. I.e. keep the type as narrow as possible in each location (as long as it doesn't lead to unreasonable verbosity, which it doesn't in this case).

Reply:
Yeah, I agree with this. Had a lapse of judgement.